File: PriorityQueue/array_based/python/unsorted_priority_queue.py
import typing

E = typing.TypeVar("E")
Number = typing.Union[int, float]


class Node(typing.Generic[E]):

    def __init__(
            self,
            element: E,
            key: Number,
            next_: typing.Optional["Node[E]"] = None,
            previous: typing.Optional["Node[E]"] = None
    ) -> None:
        self._element = element
        self._key = key
        self._next = next_
        self._previous = previous

    @property
    def next(self) -> typing.Union["Node[E]", None]:
        return self._next

    @next.setter
    def next(self, node: "Node[E]") -> None:
        self._next = node
        if node is not None:
            node._previous = self

    @property
    def previous(self) -> typing.Union["Node[E]", None]:
        return self._previous

    @previous.setter
    def previous(self, node: "Node[E]") -> None:
        self._previous = node
        if node is not None:
            node._next = self

    @property
    def element(self) -> E:
        return self._element

    @element.setter
    def element(self, element: E) -> None:
        self._element = element

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}(element={self._element}, key={self._key})"

    def __lt__(self, other: "Node[E]") -> bool:
        return self._key < other._key

    def __gt__(self, other: "Node[E]") -> bool:
        return self._key > other._key


class DoublyLinkedList(typing.Generic[E]):

    def __init__(
            self,
            head: typing.Optional[Node[E]] = None,
            tail: typing.Optional[Node[E]] = None
    ) -> None:
        self._head = head
        self._tail = tail
        self._size = 0
        self._next: typing.Union[Node[E], None] = None

    @property
    def size(self) -> int:
        return self._size

    @property
    def head(self) -> typing.Union[Node[E], None]:
        return self._head

    @head.setter
    def head(self, node: Node[E]) -> None:
        if self._head is not None:
            self._head.previous = node
        self._head = node

    @property
    def tail(self) -> typing.Union[Node[E], None]:
        return self._tail

    @tail.setter
    def tail(self, node: Node[E]) -> None:
        if self._tail is not None:
            self._tail.next = node
        self._tail = node

    @property
    def empty(self) -> bool:
        return self._size < 1

    def append(self, node: Node[E]) -> None:
        self.tail = node
        if self._head is None:
            self.head = node
        self._size += 1

    def prepend(self, node: Node[E]) -> None:
        self.head = node
        if self._tail is None:
            self.tail = node
        self._size += 1

    def remove(self, node: Node[E]) -> None:
        next_ = node.next
        previous = node.previous
        if next_ is not None:
            next_.previous = previous
            if node == self._head:
                self._head = next_
        elif previous is not None:
            previous.next = None
            if node == self._tail:
                self._tail = previous
        node.next = None
        node.previous = None
        self._size -= 1

    def __iter__(self) -> typing.Iterator[Node[E]]:
        self._next = self._head
        return self

    def __next__(self) -> Node[E]:
        if self._next is None:
            raise StopIteration
        n = self._next
        self._next = n.next
        return n


class PriorityQueue(typing.Generic[E]):

    def __init__(self) -> None:
        self._list: DoublyLinkedList[E] = DoublyLinkedList()

    @property
    def size(self) -> int:
        return self._list.size

    @property
    def empty(self) -> bool:
        return self._list.empty

    def insert(self, key: Number, element: E) -> Node[E]:
        node = Node(element, key)
        self._list.append(node)
        return node

    def min(self) -> typing.Union[Node[E], None]:
        if self.empty:
            return None
        min_node = self._list.head
        for node in self._list:
            if node < min_node:
                min_node = node
        return min_node

    def remove_min(self) -> typing.Union[Node[E], None]:
        if self.empty:
            return None
        node = self.min()
        self._list.remove(node)
        return node


if __name__ == "__main__":
    queue: PriorityQueue[int] = PriorityQueue()
    queue.insert(5, 23)
    queue.insert(1, 53)
    queue.insert(0, 324)
    print(queue.size)
    print(queue.min())

File: PriorityQueue/java/src/arraybased/UnSortedPriorityQueue.java
package arraybased;

import java.util.Iterator;
import java.util.NoSuchElementException;

public class UnSortedPriorityQueue<E> {
    private final DoublyLinkedList<E> list;

    public UnSortedPriorityQueue() {
        this.list = new DoublyLinkedList<>();
    }

    public int size() {
        return list.size();
    }

    public boolean isEmpty() {
        return list.isEmpty();
    }

    public Node<E> insert(Number key, E element) {
        Node<E> node = new Node<>(element, key);
        list.append(node);
        return node;
    }

    public Node<E> min() {
        if (isEmpty())
            return null;
        Node<E> minNode = list.getHead();
        for (Node<E> node : list) {
            if (node.compareTo(minNode) < 0)
                minNode = node;
        }
        return minNode;
    }

    public Node<E> removeMin() {
        if (isEmpty())
            return null;
        Node<E> node = min();
        list.remove(node);
        return node;
    }

    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("{");
        for (Node<E> node : list) {
            builder.append(node);
            if (node.getNext() != null)
                builder.append(", ");
        }
        builder.append("}");
        return builder.toString();
    }

    public static class Node<E> implements Comparable<Node<E>> {
        private E element;
        private final Number key;
        private Node<E> next;
        private Node<E> previous;

        public Node(E element, Number key) {
            this.element = element;
            this.key = key;
        }

        public E getElement() {
            return element;
        }

        public void setElement(E element) {
            this.element = element;
        }

        public Node<E> getNext() {
            return next;
        }

        public void setNext(Node<E> node) {
            this.next = node;
            if (node == null)
                return;
            node.previous = this;
        }

        public Node<E> getPrevious() {
            return previous;
        }

        public void setPrevious(Node<E> node) {
            this.previous = node;
            if (node == null)
                return;
            node.next = this;
        }

        @SuppressWarnings({"unchecked", "rawtypes"})
        public int compareTo(Node<E> node) {
            return ((Comparable) key).compareTo(node.key);
        };

        public String toString() {
            return String.format("%s(element=%s, key=%s)", getClass().getName(), element, key);
        }
    }

    private static class DoublyLinkedList<E> implements Iterable<Node<E>> {
        private Node<E> head;
        private Node<E> tail;
        private int size = 0;

        public DoublyLinkedList(Node<E> head, Node<E> tail) {
            this.head = head;
            this.tail = tail;
        }

        public DoublyLinkedList(Node<E> head) {
            this(head, null);
        }

        public DoublyLinkedList() {
            this(null);
        }

        public void setHead(Node<E> node) {
            if (head != null) {
                head.setPrevious(node);
            }
            this.head = node;
        }

        public Node<E> getHead() {
            return head;
        }

        public void setTail(Node<E> node) {
            if (tail != null) {
                tail.setNext(node);
            }
            this.tail = node;
        }

        public Node<E> getTail() {
            return tail;
        }

        public int size() {
            return size;
        }

        public boolean isEmpty() {
            return size < 1;
        }

        public void append(Node<E> node) {
            setTail(node);
            if (head == null)
                setHead(node);
            size++;
        }

        public void append(E element, Number key) {
            append(new Node<>(element, key));
        }

        public void prepend(Node<E> node) {
            setHead(node);
            if (tail == null)
                setTail(node);
            size++;
        }

        public void prepend(E element, Number key) {
            prepend(new Node<>(element, key));
        }

        public void remove(Node<E> node) {
            Node<E> next = node.getNext();
            Node<E> previous = node.getPrevious();
            if (next != null) {
                next.setPrevious(previous);
                if (node == head)
                    this.head = next;
            } else if (previous != null) {
                previous.setNext(null);
                if (node == tail)
                    this.tail = previous;
            }
            node.setNext(null);
            node.setPrevious(null);
            size--;
        }

        @Override
        public Iterator<Node<E>> iterator() {
            return new DoublyLinkedListIterator();
        }

        private class DoublyLinkedListIterator implements Iterator<Node<E>> {
            private Node<E> next;

            public DoublyLinkedListIterator() {
                this.next = DoublyLinkedList.this.head;
            }

            @Override
            public boolean hasNext() {
                return next != null;
            }

            @Override
            public Node<E> next() {
                if (!hasNext()) {
                    throw new NoSuchElementException();
                }
                Node<E> n = next;
                this.next = n.getNext();
                return n;
            }
        }
    }

    public static void main(String[] args) {
        UnSortedPriorityQueue<Integer> queue = new UnSortedPriorityQueue<>();
        queue.insert(5, 23);
        queue.insert(1, 53);
        queue.insert(0, 324);
        System.out.println(queue.size());
        System.out.println(queue.min());
        System.out.println(queue);
    }
}
